### PR TITLE
fs: allow to configure per-alloaction heap overhead

### LINKED ISSUE #36962: littlefs: Too small heap for file cache (again).
The issue [#31524](https://github.com/zephyrproject-rtos/zephyr/issues/31524) is back.
Tested with Zephyr at b61ea62b6f09136e08f7fe7cf8a1012b5c5042b3.

This time defining FC_HEAP_PER_ALLOC_OVERHEAD as 28 make problem disappear (at least for 32 bit ARM platform).

Maybe in case when CONFIG_FS_LITTLEFS_FC_HEAP_SIZE is not postive it would be reasonable to use slab instead of heap i.e.:

``` C
#if (CONFIG_FS_LITTLEFS_FC_HEAP_SIZE - 0) > 0
static K_HEAP_DEFINE(file_cache_heap, CONFIG_FS_LITTLEFS_FC_HEAP_SIZE);
#else
#undef CONFIG_FS_LITTLEFS_FC_HEAP_SIZE
static K_MEM_SLAB_DEFINE(file_cache_pool, CONFIG_FS_LITTLEFS_CACHE_SIZE,
                CONFIG_FS_LITTLEFS_NUM_FILES, 4);
#endif /* CONFIG_FS_LITTLEFS_FC_HEAP_SIZE */


static inline void *fc_allocate(size_t size)
{
	void *ret = NULL;

#if defined(CONFIG_FS_LITTLEFS_FC_HEAP_SIZE)
	ret = k_heap_alloc(&file_cache_heap, size, K_NO_WAIT);
#else
	(void) size;
	k_mem_slab_alloc(&file_cache_pool, &ret, K_NO_WAIT);
#endif

	return ret;
}

static inline void fc_release(void *buf)
{
#if defined(CONFIG_FS_LITTLEFS_FC_HEAP_SIZE)
	k_heap_free(&file_cache_heap, buf);
#else
	k_mem_slab_free(&file_cache_pool, &buf);
#endif
}
```

### PR DESCRIPTION
**Pull request by r2r0**
The size of overhead for each heap allocation can change after
heap implementation change and such change impacts automatic
calculation of heap size for littleFS.
This patch allows per-alloaction overhead to be configurable and then
This is a temporary fix (work around) until per-alloaction overhead value
will be available from kernel internals.

Fixes #36962

Signed-off-by: Artur Lipowski <Artur.Lipowski@hidglobal.com>

### CODE REVIEW

File: subsys/fs/Kconfig.littlefs [hunk 1]
```diff
@@ -89,4 +89,17 @@ config FS_LITTLEFS_FC_HEAP_SIZE
...
+config FS_LITTLEFS_HEAP_PER_ALLOC_OVERHEAD_SIZE
+	int "Size of per-allocation overhead for littleFS heap in bytes"
+	default 32
+	help
+	  In case when total size of littleFS heap is automatically calculated
+	  we need to take into account overhead caused per block allocation.
+	  For the purpose of heap size calculation the size of each cache block
+	  will be increased by this value.
+
+endif # FS_LITTLEFS_FC_HEAP_SIZE == 0
```
Comment (de-nordic) [inline subsys/fs/Kconfig.littlefs:103]: The condition at line 92 differs.

File: subsys/fs/littlefs_fs.c [hunk 2]
```diff
@@ -37,18 +37,12 @@ static K_MEM_SLAB_DEFINE(file_data_pool, sizeof(struct lfs_file_data),
 static K_MEM_SLAB_DEFINE(lfs_dir_pool, sizeof(struct lfs_dir),
 			 CONFIG_FS_LITTLEFS_NUM_DIRS, 4);
 
-/* Inferred overhead, in bytes, for each k_heap_aligned allocation for
- * the filecache heap.  This relates to the CHUNK_UNIT parameter in
- * the heap implementation, but that value is not visible outside the
- * kernel.
- */
-#define FC_HEAP_PER_ALLOC_OVERHEAD 24U
```
Comment (de-nordic) [inline subsys/fs/littlefs_fs.c:45]: Leave this part, just add code here that will be setting `FC_HEAP_PER_ALLOC_OVERHEAD` to the `CONFIG_FS_LITTLEFS_HEAP_PER_ALLOC_OVERHEAD_SIZE`.
Change the description to have some TODO, FIXME to notify that this thing should be rather taken from the Kernel internals than set by user, but we do not have a way to do so now.

File: subsys/fs/Kconfig.littlefs [hunk 3]
```diff
@@ -89,4 +89,17 @@ config FS_LITTLEFS_FC_HEAP_SIZE
...
+if FS_LITTLEFS_FC_HEAP_SIZE <= 0
+
+config FS_LITTLEFS_HEAP_PER_ALLOC_OVERHEAD_SIZE
+	int "Size of per-allocation overhead for littleFS heap in bytes"
+	default 32
+	help
+	  In case when total size of littleFS heap is automatically calculated
+	  we need to take into account overhead caused per block allocation.
+	  For the purpose of heap size calculation the size of each cache block
+	  will be increased by this value.
```
Comment (de-nordic) [inline subsys/fs/Kconfig.littlefs:101]: Add note here, something like:
```NOTE: when your app fails to open pre-defined number of files, as set by FS_LITTLEFS_NUM_FILES, try to increase the value.```

**Review by de-nordic — CHANGES_REQUESTED** [review]
Change the
```This is a (temporary) fix for #36962```
in commit description to
```Fixes #36962```
And just mention somewhere in description that this is work around.

**Comment by r2r0** [comment]
@de-nordic All requested changes already addressed.

**Review by de-nordic — APPROVED** [review]
Looks OK.

**Comment by de-nordic** [comment]
@cfriedt Do you want to consider this for 2.7 LTS? This provided workaround for the  https://github.com/zephyrproject-rtos/zephyr/issues/36962

**Comment by cfriedt** [comment]
> @cfriedt Do you want to consider this for 2.7 LTS? This provided workaround for the  https://github.com/zephyrproject-rtos/zephyr/issues/36962

I would include every fix if I could but we are limited to blockers. It will get into 2.7.1 though.

**Comment by de-nordic** [comment]
@r2r0 Can you rebase this and force push to re-trigger checks?

**Review by cfriedt — APPROVED** [review]

**Pull request merged**